### PR TITLE
Migrate everything from Python 2.7 to Python 3.6

### PR DESCRIPTION
**Pull request by bkgoksel**
As pip has recently told us in all caps:

`DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7`

This is the beginning of a long, uncertain journey that we nevertheless must take.
The first step is to run `2to3` on our entire codebase and hope that the world doesn't burn around us.

Godspeed,

### CODE REVIEW

**Comment by bkgoksel** [comment]
Known issue:
Python3 doesn't have the `err.message` attribute for exceptions, need to just pass the error itself as its string representation wherever this happens ie `err.message` -> `str(err)`

File: test_cli.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def run_command(args, expected_exit_code=0, max_output_chars=256, env=None):
     for a in args:
         assert isinstance(a, str)
```
Comment (epicfaace) [inline test_cli.py]:
```suggestion
        pass
```

**Comment by epicfaace** [comment]
I kept getting this issue (when running `worksheets` tests):

```
>> codalab/bin/cl wedit 0xf51f5551139042c78fa3534dc6e19187 --file /home/epicfaace/codalab-worksheets/tests/files/unicode-worksheet
Traceback (most recent call last):
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 16, in wrapper
    return f(*args, **kwargs)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 680, in update_worksheet_raw
    data='\n'.join(lines).encode(),
  File "/home/epicfaace/codalab-worksheets/worker/codalabworker/rest_client.py", line 91, in _make_request
    with closing(urllib.request.urlopen(request)) as response:
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 500: Internal Server Error
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "codalab/bin/../../codalab/bin/cl.py", line 125, in <module>
    main()
  File "codalab/bin/../../codalab/bin/cl.py", line 119, in main
    cli.do_command(sys.argv[1:])
  File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 848, in do_command
    structured_result = command_fn()
  File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 842, in <lambda>
    command_fn = lambda: args.function(self, args)
  File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 2932, in do_wedit_command
    client.update_worksheet_raw(worksheet_info['id'], lines)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 38, in wrapper
    sys.exc_info()[2],
  File "/home/epicfaace/codalab-worksheets/venv/lib/python3.6/site-packages/six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 16, in wrapper
    return f(*args, **kwargs)
  File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 680, in update_worksheet_raw
    data='\n'.join(lines).encode(),
  File "/home/epicfaace/codalab-worksheets/worker/codalabworker/rest_client.py", line 91, in _make_request
    with closing(urllib.request.urlopen(request)) as response:
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
codalab.client.json_api_client.JsonApiException: Unable to update worksheet: Internal Server Error - Unexpected Internal Error ('latin-1' codec can't encode characters in position 7-19: ordinal not in range(256)). The administrators have been notified.
 (exit code 1, expected 0)
```
It appears that 908ce5c224ba75105aef1ceea0e73613bf812bd3 fixed it, although I'm not sure whether this is the best solution and whether this will work in production -- will need to test.

File: codalab/client/json_api_client.py [hunk 2]
```diff
@@ -186,9 +188,9 @@ def _pack_params(params):
         if include is not None:
             result.append(('include', ','.join(include)))
 
-        for k, v in params.iteritems() if isinstance(params, dict) else params:
+        for k, v in iter(list(params.items())) if isinstance(params, dict) else params:
```
Comment (epicfaace) [inline codalab/client/json_api_client.py]:
```suggestion
        for k, v in params.items() if isinstance(params, dict) else params:
```

File: codalab/client/json_api_client.py [hunk 3]
```diff
@@ -286,7 +288,7 @@ def unpack_object(obj_data):
                 obj.update(obj_data['attributes'])
             if 'meta' in obj_data:
                 obj['meta'] = obj_data['meta']
-            for key, relationship in obj_data.get('relationships', {}).iteritems():
+            for key, relationship in list(obj_data.get('relationships', {}).items()):
```
Comment (epicfaace) [inline codalab/client/json_api_client.py]:
```suggestion
            for key, relationship in obj_data.get('relationships', {}).items():
```

File: tests/worker/file_util_test.py [hunk 4]
```diff
@@ -39,26 +39,26 @@ def test_tar_empty(self):
...
         output_dir = os.path.join(temp_dir, 'output')
         un_tar_directory(tar_gzip_directory(dir), output_dir, 'gz')
-        self.assertEquals(os.listdir(output_dir), [])
+        self.assertEqual(os.listdir(output_dir), [])
 
     def test_gzip_stream(self):
         with tempfile.NamedTemporaryFile(delete=False) as temp_file:
             self.addCleanup(lambda: os.remove(temp_file.name))
-            temp_file.write('contents')
+            temp_file.write('contents'.encode())
```
Comment (epicfaace) [inline tests/worker/file_util_test.py]:
```suggestion
            temp_file.write(b'contents')
```

File: tests/worker/file_util_test.py [hunk 5]
```diff
@@ -39,26 +39,26 @@ def test_tar_empty(self):
...
             name = temp_file.name
 
-        self.assertEquals(un_gzip_stream(gzip_file(name)).read(), 'contents')
+        self.assertEqual(un_gzip_stream(gzip_file(name)).read(), b'contents')
 
     def test_bz2_file(self):
         source_write = tempfile.NamedTemporaryFile(delete=False)
         self.addCleanup(lambda: os.remove(source_write.name))
-        source_write.write(bz2.compress('contents'))
+        source_write.write(bz2.compress('contents'.encode()))
```
Comment (epicfaace) [inline tests/worker/file_util_test.py]:
```suggestion
        source_write.write(bz2.compress(b'contents'))
```

File: worker/codalabworker/bundle_service_client.py [hunk 6]
```diff
@@ -21,15 +20,18 @@ def wrapper(*args, **kwargs):
...
             except RestClientException as e:
-                raise BundleServiceException(message + ': ' + e.message, e.client_error)
-            except urllib2.HTTPError as e:
-                client_error = json.load(e)
+                raise BundleServiceException(message + ': ' + str(e), e.client_error)
+            except urllib.error.HTTPError as e:
+                try:
+                    client_error = json.loads(e.read().decode())
+                except json.JSONDecodeError:
+                    client_error = {'error': e.read().decode()}
```
Comment (epicfaace) [inline worker/codalabworker/bundle_service_client.py]: Do we need this try-catch?

File: worker/codalabworker/file_util.py [hunk 7]
```diff
@@ -104,7 +104,7 @@ class UnGzipStream(object):
         def __init__(self, fileobj):
             self._fileobj = fileobj
             self._decoder = zlib.decompressobj(16 + zlib.MAX_WBITS)
-            self._buffer = ''
+            self._buffer = ''.encode()
```
Comment (epicfaace) [inline worker/codalabworker/file_util.py]:
```suggestion
            self._buffer = b''
```

File: worker/codalabworker/main.py [hunk 8]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python3.6
```
Comment (epicfaace) [inline worker/codalabworker/main.py:1]: Should we say this or python3?

File: worker/codalabworker/main.py [hunk 9]
```diff
@@ -112,19 +112,23 @@ def main():
...
+            )
             exit(1)
         with open(args.password_file) as f:
             username = f.readline().strip()
             password = f.readline().strip()
     else:
         username = os.environ.get('CODALAB_USERNAME')
         if username is None:
-            username = raw_input('Username: ')
+            username = eval(input('Username: '))
```
Comment (epicfaace) [inline worker/codalabworker/main.py]: 2to3 replaced raw_input with eval(input(...)), but perhaps it's safer if we just keep it at input(...)?

File: worker/codalabworker/rest_client.py [hunk 10]
```diff
@@ -138,22 +136,24 @@ def _upload_with_chunked_encoding(
                 to_send = fileobj.read(CHUNK_SIZE)
                 if not to_send:
                     break
-                conn.send('%X\r\n%s\r\n' % (len(to_send), to_send))
+                if type(to_send) is str:
+                    to_send = to_send.encode()
+                conn.send(b'%X\r\n%s\r\n' % (len(to_send), to_send))
```
Comment (epicfaace) [inline worker/codalabworker/rest_client.py:139]: Previously (in Python 2) the code said:
```
conn.send('%X\r\n%s\r\n' % (len(to_send), to_send))
```

In that case, if `to_send` were a character such as `é` (which has unicode representation `\xc3\xa9`), then what would be sent in the connection would be the bytestring `1\r\n\xc3\xa9\r\n`.

However, the replacement Python 3 code I've added in would create the bytestring `2\r\n\xc3\xa9\r\n`. I'd think the new code I'm writing might be correct, because `é` is 2 bytes, not one byte, long -- but I don't know what the first number is used for later. Which is correct?

File: worker/codalabworker/worker.py [hunk 11]
```diff
@@ -152,7 +157,7 @@ def reply(err, message={}, data=None):
             traceback.print_exc()
         except Exception as e:
             traceback.print_exc()
-            err = (httplib.INTERNAL_SERVER_ERROR, e.message)
+            err = (http.client.INTERNAL_SERVER_ERROR, str(e) + "\n" + traceback.format_exc())
```
Comment (epicfaace) [inline worker/codalabworker/worker.py]:
```suggestion
            err = (http.client.INTERNAL_SERVER_ERROR, str(e))
```

File: setup.py [hunk 12]
```diff
@@ -40,7 +40,7 @@ def get_requirements(*requirements_file_paths):
         "Programming Language :: Python :: 2.7",
         "License :: OSI Approved :: Apache Software License",
     ],
-    python_requires='~=2.7',
+    python_requires='~=3.0',
```
Comment (epicfaace) [inline setup.py]:
```suggestion
    python_requires='~=3.5',
```

File: worker/setup.py [hunk 13]
```diff
@@ -27,11 +27,11 @@ def get_requirements(*requirements_file_paths):
...
     packages=find_packages(include=['codalabworker*']),
     classifiers=[
-        "Programming Language :: Python :: 2 :: Only",
-        "Programming Language :: Python :: 2.7",
+        "Programming Language :: Python :: 3 :: Only",
+        "Programming Language :: Python :: 3.6",
         "License :: OSI Approved :: Apache Software License",
     ],
-    python_requires='~=2.7',
+    python_requires='~=3.0',
```
Comment (epicfaace) [inline worker/setup.py]:
```suggestion
    python_requires='~=3.5',
```

File: codalab/client/json_api_client.py [hunk 14]
```diff
@@ -14,9 +13,11 @@ def decorator(f):
         def wrapper(*args, **kwargs):
             try:
                 return f(*args, **kwargs)
-            except urllib2.HTTPError as e:
+            except urllib.error.HTTPError as e:
                 # Translate known errors to the standard CodaLab errors
                 error_body = e.read()
+                if type(error_body) is bytes:
```
Comment (epicfaace) [inline codalab/client/json_api_client.py]: Todo: do we need this?

File: codalab/model/bundle_model.py [hunk 15]
```diff
@@ -1074,20 +1074,20 @@ def batch_get_worksheets(self, fetch_items, **kwargs):
...
+            for value in worksheet_values.values():
                 value['items'] = []
             for item_row in sorted(item_rows, key=item_sort_key):
                 if item_row.worksheet_uuid not in worksheet_values:
                     raise IntegrityError('Got item %s without worksheet' % (item_row,))
                 item_row = dict(item_row)
                 item_row['value'] = self.decode_str(item_row['value'])
                 worksheet_values[item_row['worksheet_uuid']]['items'].append(item_row)
-        return [Worksheet(value) for value in worksheet_values.itervalues()]
+        return [Worksheet(value) for value in list(worksheet_values.values())]
```
Comment (epicfaace) [inline codalab/model/bundle_model.py]:
```suggestion
        return [Worksheet(value) for value in worksheet_values.values()]
```

File: codalab/model/bundle_model.py [hunk 16]
```diff
@@ -1502,7 +1502,7 @@ def batch_get_groups(self, **kwargs):
             if not rows:
                 return []
         values = {row.uuid: str_key_dict(row) for row in rows}
-        return [value for value in values.itervalues()]
+        return [value for value in list(values.values())]
```
Comment (epicfaace) [inline codalab/model/bundle_model.py]:
```suggestion
        return [value for value in values.values()]
```

File: codalab/model/bundle_model.py [hunk 17]
```diff
@@ -1568,7 +1568,7 @@ def batch_get_all_groups(self, spec_filters, group_filters, user_group_filters):
                     row['owner_id'] = str(row['owner_id'])
                 rows[i] = row
             values = {row['uuid']: row for row in rows}
-            return [value for value in values.itervalues()]
+            return [value for value in list(values.values())]
```
Comment (epicfaace) [inline codalab/model/bundle_model.py]:
```suggestion
            return [value for value in values.values()]
```

File: codalab/model/mysql_model.py [hunk 18]
```diff
@@ -60,7 +60,7 @@ def do_multirow_insert(self, connection, table, values):
     # (See tables.py for more details.)
 
     def encode_str(self, value):
-        return value.encode('utf-8')
+        return value
```
Comment (epicfaace) [inline codalab/model/mysql_model.py]: This _shouldn't_ break anything, I think...

File: codalab/model/worker_model.py [hunk 19]
```diff
@@ -281,7 +281,10 @@ def get_json_message(self, sock, timeout_secs):
             return None
 
         with closing(fileobj):
-            return json.loads(fileobj.read())
+            try:
+                return json.loads(fileobj.read().decode())
+            except json.JSONDecodeError:
```
Comment (epicfaace) [inline codalab/model/worker_model.py]: Do we need this try-except?

File: codalab/objects/permission.py [hunk 20]
```diff
@@ -104,7 +104,7 @@ def _check_permissions(model, table, user, object_uuids, owner_ids, need_permiss
             user_str,
             object_type,
             ' '.join(object_uuids),
-            ' '.join(map(permission_str, have_permissions.values())),
+            ' '.join(map(permission_str, list(have_permissions.values()))),
```
Comment (epicfaace) [inline codalab/objects/permission.py:107]:
```suggestion
            ' '.join(map(permission_str, have_permissions.values())),
```

File: codalab/rest/util.py [hunk 21]
```diff
@@ -100,14 +100,14 @@ def get_bundle_infos(
         parent2children = local.model.get_children_uuids(readable)
 
         # Gather all children bundle uuids and fetch permissions
-        child_uuids = [uuid for l in parent2children.values() for uuid in l]
+        child_uuids = [uuid for l in list(parent2children.values()) for uuid in l]
```
Comment (epicfaace) [inline codalab/rest/util.py]:
```suggestion
        child_uuids = [uuid for l in parent2children.values() for uuid in l]
```

File: codalab/rest/util.py [hunk 22]
```diff
@@ -118,15 +118,15 @@ def get_bundle_infos(
         # bundle_uuids -> list of worksheet_uuids
         host_worksheets = local.model.get_host_worksheet_uuids(readable)
         # Gather all worksheet uuids
-        worksheet_uuids = [uuid for l in host_worksheets.itervalues() for uuid in l]
+        worksheet_uuids = [uuid for l in list(host_worksheets.values()) for uuid in l]
```
Comment (epicfaace) [inline codalab/rest/util.py]:
```suggestion
        worksheet_uuids = [uuid for l in host_worksheets.values() for uuid in l]
```

File: codalab_service.py [hunk 23]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/python2.7
+#! /usr/bin/python3
```
Comment (epicfaace) [inline codalab_service.py]: Should this be python3.6?

File: test_cli.py [hunk 24]
```diff
@@ -101,25 +101,22 @@ def get_uuid(line):
...
 
 def sanitize(string, max_chars=256):
-    try:
-        string.decode('utf-8')
-    except UnicodeDecodeError:
-        return '<binary>\n'
-
-    string = string.decode('utf-8').encode('ascii', errors='replace')  # Travis only prints ASCII
+    if type(string) is bytes:
+        string = "{}".format(string)
```
Comment (epicfaace) [inline test_cli.py:109]: For some reason this shows up as `<binary>` on the old travis CI runs, but on the new travis CI runs it ends up showing the entire (truncated) contents of the binary file. Do we know why?

File: test_cli.py [hunk 25]
```diff
@@ -518,7 +519,7 @@ def test(ctx):
     check_contains(['c', 'd', 'e'], get_info(uuid, 'tags'))
 
     # cat, info
-    check_equals(test_path_contents('a.txt'), run_command([cl, 'cat', uuid]))
+    check_equals(test_path_contents('a.txt'), run_command([cl, 'cat', uuid]))  # here?
```
Comment (epicfaace) [inline test_cli.py]:
```suggestion
    check_equals(test_path_contents('a.txt'), run_command([cl, 'cat', uuid]))
```

File: worker/codalabworker/local_run/local_dependency_manager.py [hunk 26]
```diff
@@ -185,7 +185,7 @@ def _cleanup(self):
         while True:
             with self._global_lock:
                 self._acquire_all_locks()
-                bytes_used = sum(dep.size_bytes for dep in self._dependencies.values())
+                bytes_used = sum(dep.size_bytes for dep in list(self._dependencies.values()))
```
Comment (epicfaace) [inline worker/codalabworker/local_run/local_dependency_manager.py]:
```suggestion
                bytes_used = sum(dep.size_bytes for dep in self._dependencies.values())
```

File: worker/codalabworker/local_run/local_run_manager.py [hunk 27]
```diff
@@ -173,7 +173,7 @@ def process_runs(self):
             # filter out finished runs
             finished_container_ids = [
                 run.container
-                for run in self._runs.values()
+                for run in list(self._runs.values())
```
Comment (epicfaace) [inline worker/codalabworker/local_run/local_run_manager.py]:
```suggestion
                for run in self._runs.values()
```

File: codalab/client/json_api_client.py (hunk 3 above)
Comment (epicfaace) [reply codalab/client/json_api_client.py]:
```suggestion
            for key, relationship in obj_data.get('relationships', {}).items():
```

File: codalab/client/json_api_client.py (hunk 14 above)
Comment (percyliang) [reply codalab/client/json_api_client.py]: does `e.read()` not return a consistent type?

File: codalab/lib/codalab_manager.py [hunk 28]
```diff
@@ -190,7 +190,7 @@ def init_config(self, dry_run=False):
                 'rest_host': 'localhost',
                 'rest_port': 2900,
                 'class': 'MySQLModel',
-                'engine_url': 'mysql://codalab@localhost:3306/codalab_bundles',
+                'engine_url': 'mysql://codalab@localhost:3306/codalab_bundles?charset=utf8',
```
Comment (percyliang) [inline codalab/lib/codalab_manager.py]: Is this necessary?  Do all mysql URLs need this now?  If people upgrade, their local file will not change.  There might be other places where we need to update this?

File: codalab/lib/codalab_manager.py [hunk 29]
```diff
@@ -346,7 +346,8 @@ def model(self):
             # Patch for backwards-compatibility until we have a cleaner abstraction around config
             # that can update configs to newer "versions"
             engine_url = self.config['server'].get(
-                'engine_url', "sqlite:///{}".format(os.path.join(self.codalab_home, 'bundle.db'))
+                'engine_url',
+                "sqlite:///{}?charset=utf8".format(os.path.join(self.codalab_home, 'bundle.db')),
```
Comment (percyliang) [inline codalab/lib/codalab_manager.py]: Same with this.

File: codalab/lib/download_manager.py [hunk 30]
```diff
@@ -288,7 +288,7 @@ def _get_read_response_stream(self, response_socket_id):
 
     def _get_read_response_string(self, response_socket_id):
         with closing(self._get_read_response_stream(response_socket_id)) as fileobj:
-            return fileobj.read()
+            return fileobj.read().decode()
```
Comment (percyliang) [inline codalab/lib/download_manager.py:291]: Are we assuming Unicode for all files?  What about binary files?

File: docker/dockerfiles/Dockerfile.default-cpu [hunk 31]
```diff
@@ -1,17 +1,19 @@
...
 
 RUN apt-get update && apt-get install -y \
     apt-transport-https \
     iputils-ping \
     git \
-    python2.7 \
-    python-pip \
-    python-dev \
-    python-software-properties \
+    python2.7 \
```
Comment (percyliang) [inline docker/dockerfiles/Dockerfile.default-cpu:13]: Don't need 2.7 anymore right?

File: docker/dockerfiles/Dockerfile.default-cpu [hunk 32]
```diff
@@ -75,8 +77,8 @@ RUN python -m spacy download en
 
 ## Python 3 packages
 
-RUN pip3 install -U pip
-RUN pip3 install -U \
+RUN python3.6 -m pip install -U pip
```
Comment (percyliang) [inline docker/dockerfiles/Dockerfile.default-cpu]: Does pip3 not work anymore?

File: docker/dockerfiles/Dockerfile.default-gpu [hunk 33]
```diff
@@ -8,10 +8,10 @@ RUN apt-get update && apt-get install -y \
     apt-transport-https \
     iputils-ping \
     git \
-    python2.7 \
-    python-pip \
-    python-dev \
-    python-software-properties \
+    python2.7 \
```
Comment (percyliang) [inline docker/dockerfiles/Dockerfile.default-gpu:11]: Remove trailing whitespace

File: test_cli.py (hunk 24 above)
Comment (percyliang) [reply test_cli.py:109]: Can we avoid this condition on type pattern?  I think it makes it a bit hard to reason about what types are expected where.

**Comment by epicfaace** [comment]
@percyliang did you try this out yourself -- do you think it's ready to merge?

File: worker/codalabworker/bundle_service_client.py (hunk 6 above)
Comment (percyliang) [inline worker/codalabworker/bundle_service_client.py]: why are we decoding these strings?

File: worker/codalabworker/main.py (hunk 8 above)
Comment (percyliang) [reply worker/codalabworker/main.py:1]: python3

File: worker/codalabworker/rest_client.py [hunk 34]
```diff
@@ -63,37 +61,37 @@ def _make_request(
...
-            path = path + '?' + urllib.urlencode(query_params)
+            path = path + '?' + urllib.parse.urlencode(query_params)
 
-        # Everything needs to be utf-8 encoded or else urllib2 will complain
+        # Everything needs to be utf-8 encoded or else urllib will complain
         if 'Content-Type' in headers:
             headers['Content-Type'] += '; charset=utf-8'
-        if data and isinstance(data, unicode):
+        if isinstance(data, str):
             data = data.encode('utf-8')
```
Comment (percyliang) [inline worker/codalabworker/rest_client.py]: do we need 'utf-8' explicitly?

**Review by percyliang — COMMENTED** [review]
Great to see this PR coming together - it will be so much nicer to be on Python 3!

I still need to try out the code and test drive it more thoroughly - since this is a big change, I'd want to make sure everything's okay.

Concretely, the main thing I'm concerned about is the handling of strings and unicode, which has been a huge issue with the existing code.  With Python 3 and everything being unicode, things should be much nicer, but there are still a bunch of byte-string conversions that exist at various points in the code.  Could you summarize briefly what the contract is for where things are bytes and where things are strings and where conversions need to happen?

File: docker/dockerfiles/Dockerfile.default-cpu (hunk 31 above)
Comment (epicfaace) [reply docker/dockerfiles/Dockerfile.default-cpu:13]: Don't we need 2.7 so that the workers can run scripts in Python 2?

File: worker/codalabworker/rest_client.py (hunk 10 above)
Comment (epicfaace) [reply worker/codalabworker/rest_client.py:139]: @percyliang

File: codalab/lib/codalab_manager.py (hunk 28 above)
Comment (epicfaace) [reply codalab/lib/codalab_manager.py]: > I kept getting this issue (when running `worksheets` tests):
> 
> ```
> >> codalab/bin/cl wedit 0xf51f5551139042c78fa3534dc6e19187 --file /home/epicfaace/codalab-worksheets/tests/files/unicode-worksheet
> Traceback (most recent call last):
>   File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 16, in wrapper
>     return f(*args, **kwargs)
>   File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 680, in update_worksheet_raw
>     data='\n'.join(lines).encode(),
>   File "/home/epicfaace/codalab-worksheets/worker/codalabworker/rest_client.py", line 91, in _make_request
>     with closing(urllib.request.urlopen(request)) as response:
>   File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib/python3.6/urllib/request.py", line 532, in open
>     response = meth(req, response)
>   File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
>     'http', request, response, code, msg, hdrs)
>   File "/usr/lib/python3.6/urllib/request.py", line 570, in error
>     return self._call_chain(*args)
>   File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> urllib.error.HTTPError: HTTP Error 500: Internal Server Error
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "codalab/bin/../../codalab/bin/cl.py", line 125, in <module>
>     main()
>   File "codalab/bin/../../codalab/bin/cl.py", line 119, in main
>     cli.do_command(sys.argv[1:])
>   File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 848, in do_command
>     structured_result = command_fn()
>   File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 842, in <lambda>
>     command_fn = lambda: args.function(self, args)
>   File "/home/epicfaace/codalab-worksheets/codalab/lib/bundle_cli.py", line 2932, in do_wedit_command
>     client.update_worksheet_raw(worksheet_info['id'], lines)
>   File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 38, in wrapper
>     sys.exc_info()[2],
>   File "/home/epicfaace/codalab-worksheets/venv/lib/python3.6/site-packages/six.py", line 692, in reraise
>     raise value.with_traceback(tb)
>   File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 16, in wrapper
>     return f(*args, **kwargs)
>   File "/home/epicfaace/codalab-worksheets/codalab/client/json_api_client.py", line 680, in update_worksheet_raw
>     data='\n'.join(lines).encode(),
>   File "/home/epicfaace/codalab-worksheets/worker/codalabworker/rest_client.py", line 91, in _make_request
>     with closing(urllib.request.urlopen(request)) as response:
>   File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib/python3.6/urllib/request.py", line 532, in open
>     response = meth(req, response)
>   File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
>     'http', request, response, code, msg, hdrs)
>   File "/usr/lib/python3.6/urllib/request.py", line 570, in error
>     return self._call_chain(*args)
>   File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
>     raise HTTPError(req.full_url, code, msg, hdrs, fp)
> codalab.client.json_api_client.JsonApiException: Unable to update worksheet: Internal Server Error - Unexpected Internal Error ('latin-1' codec can't encode characters in position 7-19: ordinal not in range(256)). The administrators have been notified.
>  (exit code 1, expected 0)
> ```
> 
> It appears that [908ce5c](https://github.com/codalab/codalab-worksheets/commit/908ce5c224ba75105aef1ceea0e73613bf812bd3) fixed it, although I'm not sure whether this is the best solution and whether this will work in production -- will need to test.

@percyliang see this comment regarding the database urls.

File: docker/dockerfiles/Dockerfile.default-cpu (hunk 31 above)
Comment (percyliang) [reply docker/dockerfiles/Dockerfile.default-cpu:13]: You're right, never mind.

**Comment by percyliang** [comment]
Just tried testing the system with unicode and found the following problems:
- If I add '你好' into the worksheet markdown (either using `cl wedit` or in the web browser), it shows up as '??'.
- If I upload a file with unicode, `cl cat` works but `summarize_file` fails with "Unexpected Internal Error ('ascii' codec can't decode byte 0xe8 in position 0: ordinal not in range(128)). The administrators have been notified." 
- `cl wedit -T 你好` throws an exception
- `cl edit <bundle> -d 你好` also saves as '??'
Probably should have tests for these cases.

**Comment by epicfaace** [comment]
@percyliang In each of the above cases, should codalab allow or not allow non-ascii characters like "你好"?

**Comment by epicfaace** [comment]
For example, it seems like a lot of fields in the schema (such as worksheet title, tags) are supposed to be ascii -- why was this this case? Do we want to change it to allow unicode? https://github.com/codalab/codalab-worksheets/blob/54ab716491b774fd66a7172120c21bea543360b3/codalab/rest/schemas.py#L119

**Comment by epicfaace** [comment]
> * If I upload a file with unicode, `cl cat` works but `summarize_file` fails with "Unexpected Internal Error ('ascii' codec can't decode byte 0xe8 in position 0: ordinal not in range(128)). The administrators have been notified."

How do you run `summarize_file`? Is there a way to run it from the command line?

**Comment by percyliang** [comment]
Ideally we should allow unicode everywhere.  Not sure why the validate_ascii was there - probably because unicode wasn't working before.

File: worker/codalabworker/rest_client.py (hunk 34 above)
Comment (epicfaace) [reply worker/codalabworker/rest_client.py]: No, changing.

Comment (epicfaace) [inline worker/codalabworker/rest_client.py]:
```suggestion
            data = data.encode()
```

File: codalab/lib/codalab_manager.py [hunk 35]
```diff
@@ -71,7 +71,7 @@ def write_pretty_json(data, path):
 def read_json_or_die(path):
     try:
         with open(path, 'rb') as f:
-            string = f.read()
+            string = f.read().decode('utf-8')
```
Comment (epicfaace) [inline codalab/lib/codalab_manager.py]: Do we need rb here?

File: codalab/lib/download_manager.py (hunk 30 above)
Comment (epicfaace) [reply codalab/lib/download_manager.py:291]: Yes, I think in this case it is always a string. If I remove `.decode()`, I get this exception when running netcat tests:

```
Traceback (most recent call last):
  File "/usr/local/bin/cl", line 11, in <module>
    load_entry_point('codalab', 'console_scripts', 'cl')()
  File "/opt/codalab-worksheets/codalab/bin/cl.py", line 120, in main
    cli.do_command([os.fsencode(arg).decode() for arg in sys.argv[1:]])
  File "/opt/codalab-worksheets/codalab/lib/bundle_cli.py", line 854, in do_command
    structured_result = command_fn()
  File "/opt/codalab-worksheets/codalab/lib/bundle_cli.py", line 848, in <lambda>
    command_fn = lambda: args.function(self, args)
  File "/opt/codalab-worksheets/codalab/lib/bundle_cli.py", line 2225, in do_netcat_command
    info = client.netcat(bundle_uuid, port=args.port, data={"message": args.message})
  File "/opt/codalab-worksheets/codalab/client/json_api_client.py", line 35, in wrapper
    sys.exc_info()[2],
  File "/usr/local/lib/python3.6/dist-packages/six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "/opt/codalab-worksheets/codalab/client/json_api_client.py", line 15, in wrapper
    return f(*args, **kwargs)
  File "/opt/codalab-worksheets/codalab/client/json_api_client.py", line 445, in netcat
    return self._make_request('PUT', request_path, data=data)
  File "/opt/codalab-worksheets/worker/codalabworker/rest_client.py", line 89, in _make_request
    with closing(urllib.request.urlopen(request)) as response:
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
codalab.client.json_api_client.JsonApiException: Unable to netcat 0x3799cc78981f420d88c72ad4de3ec05c: Internal Server Error - Unexpected Internal Error (Object of type 'bytes' is not JSON serializable). The administrators have been notified.
```

File: codalab/model/worker_model.py (hunk 19 above)
Comment (epicfaace) [reply codalab/model/worker_model.py]: No, we don't.

File: worker/codalabworker/bundle_service_client.py (hunk 6 above)
Comment (epicfaace) [reply worker/codalabworker/bundle_service_client.py]: @percyliang urllib2.HTTPError returns a file-like object which is opened in **binary** mode. Thus, `json.load` doesn't work on it, so we have to read and convert it to a string first.

Comment (epicfaace) [reply worker/codalabworker/bundle_service_client.py]: No, we don't.

File: worker/codalabworker/rest_client.py (hunk 10 above)
Comment (epicfaace) [reply worker/codalabworker/rest_client.py:139]: Resolved in #1336.

**Review by percyliang — APPROVED** [review]
This is great!  Remember to increment the version to 0.4.0.

**Comment by epicfaace** [comment]
Why 0.4.0 (as opposed to, say, 1.0.0)?

**Comment by percyliang** [comment]
I was thinking `0.4.0` since it's the same functionality and is backwards compatible...

**Comment by epicfaace** [comment]
I'll increment the version in another PR.
